plugins/processors/flux/flux.go

```diff
@@ -0,0 +1,342 @@
+	"github.com/influxdata/telegraf/plugins/processors"
+)
+
+const (
+	description  = "Runs a flux script to process inputs and produce outputs"
+	sampleConfig = `
+  ## Path to a Flux script.
+  path = "path/to/script.flux"
+  ## Enables debug mode if an output file is specified.
+  debug = "path/to/out.csv"
+`
+	measurementColumnLabel = "_measurement"
+)
+
+type Flux struct {
+	Path  string `toml:"path"`
+	Debug string `toml:"debug"`
+
+	Log telegraf.Logger `toml:"-"`
+
+	debugFile *os.File
+	program   flux.Program
+}
+
+func init() {
+	processors.Add("flux", func() telegraf.Processor {
+		return &Flux{}
+	})
+}
+
+func (p *Flux) SampleConfig() string {
+	return sampleConfig
+}
+
+func (p *Flux) Description() string {
+	return description
+}
+
+func (p *Flux) init() error {
+	if err := p.compile(); err != nil {
+		return fmt.Errorf("script failed to compile: %v", err)
+	}
+	if len(p.Debug) > 0 && p.debugFile == nil {
+		f, err := os.OpenFile(p.Debug, os.O_RDWR|os.O_CREATE|os.O_APPEND, 0666)
+		if err != nil {
+			return fmt.Errorf("could not open debug file: %s", p.Debug)
+		}
+		p.debugFile = f
+	}
+	return nil
+}
+
+func (p *Flux) Apply(metrics ...telegraf.Metric) []telegraf.Metric {
+	if len(metrics) == 0 {
+		return nil
+	}
+	if err := p.init(); err != nil {
+		p.Log.Errorf("Error on flux processor init: %v", err)
+		return nil
+	}
+
+	tbls, err := getTables(metrics...)
+	if err != nil {
+		p.Log.Errorf("Could not extract tables from metrics: %v", err)
+		return nil
+	}
+	if len(p.Debug) > 0 {
+		encoded, err := getCSV(tbls)
+		if err != nil {
+			p.Log.Errorf("Error in encoding CSV: %v", err)
+			return nil
+		}
+		if _, err := p.debugFile.WriteString(encoded); err != nil {
+			p.Log.Errorf("Error while writing to debug file: %v", err)
+		}
+		return nil
+	}
+	max := time.Unix(0, 0)
+	for _, m := range metrics {
+		if m.Time().After(max) {
+			max = m.Time()
+		}
+	}
+	provider, err := tables.NewProvider(tbls)
+	if err != nil {
+		p.Log.Errorf("Internal error while creating table provider: %v", err)
+		return nil
+	}
+	deps := Deps{
+		Tables: provider,
+		Max:    max,
+	}
+	ctx := deps.Inject(context.Background())
+	q, err := p.program.Start(ctx, new(memory.Allocator))
+	if err != nil {
+		p.Log.Errorf("Error while running program: %v", err)
+		return nil
+	}
+	defer q.Done()
+	ms, err := getMetrics(q)
+	if err != nil {
+		p.Log.Errorf("Error while running program: %v", err)
+		return nil
+	}
+	return ms
+}
+
+func (p *Flux) compile() error {
+	if p.program != nil {
+		return nil
+	}
+	bs, err := ioutil.ReadFile(p.Path)
+	if err != nil {
+		return err
+	}
+	script := string(bs)
+	astPkg, err := flux.Parse(script)
+	if err != nil {
+		return err
+	}
+	p.program = lang.CompileAST(astPkg, time.Now())
+	return nil
+}
+
+func getTables(metrics ...telegraf.Metric) ([]flux.Table, error) {
+	tbls := make([]flux.Table, 0)
+	builders := execute.NewGroupLookup()
+	getBuilder := func(gk flux.GroupKey) (execute.TableBuilder, bool) {
```

Comment:
So this function groups the metric by the given key. Is my understanding correct? Could you please spend a small comment for all non-flux devs. :-)